src/borg/archiver/prune_cmd.py

```diff
@@ -77,20 +92,36 @@ def quarterly_3monthly_period_func(a):
 
 PRUNING_PATTERNS = OrderedDict(
     [
-        ("secondly", default_period_func("%Y-%m-%d %H:%M:%S")),
-        ("minutely", default_period_func("%Y-%m-%d %H:%M")),
-        ("hourly", default_period_func("%Y-%m-%d %H")),
-        ("daily", default_period_func("%Y-%m-%d")),
-        ("weekly", default_period_func("%G-%V")),
-        ("monthly", default_period_func("%Y-%m")),
+        # Each archive is considered for keeping
+        ("within", unique_period_func()),
+        ("last", unique_period_func()),
+        ("keep", unique_period_func()),
+        # Last archive (by creation timestamp) within period group is consiedered for keeping
+        ("secondly", pattern_period_func("%Y-%m-%d %H:%M:%S")),
+        ("minutely", pattern_period_func("%Y-%m-%d %H:%M")),
+        ("hourly", pattern_period_func("%Y-%m-%d %H")),
+        ("daily", pattern_period_func("%Y-%m-%d")),
+        ("weekly", pattern_period_func("%G-%V")),
+        ("monthly", pattern_period_func("%Y-%m")),
         ("quarterly_13weekly", quarterly_13weekly_period_func),
         ("quarterly_3monthly", quarterly_3monthly_period_func),
-        ("yearly", default_period_func("%Y")),
+        ("yearly", pattern_period_func("%Y")),
     ]
 )
 
 
-def prune_split(archives, rule, n, kept_because=None):
+# Datetime cannot represent times than datetime.min, so a day is added to allow for time zone offset.
+DATETIME_MIN_WITH_ZONE = (datetime.min + timedelta(days=1)).astimezone()
```

Comment:
It doesn't work to just use datetime.min?  Is that because we can only compare timestamps in the same time zone?

Reply:
Maybe it could work with timezone set to UTC, I can try that. As long as the timestamps being compared both have zones it should be good. But taking `datetime.zone` and converting it to my local timezone (currently +02) gives an underflow error.

Reply:
Can confirm that `.astimezone(timezone.utc)` does not work, giving the same error. Again, my local timezone is not equal to UTC and it seems `datetime.min` is implicitly in the local timezone already.

Again, since (at least in testing) the timestamps from the archive have timezone info, the datetime_min timestamp must also have timezone info. We could of course go the other way and make them both native UTC timestamps, but I think it makes for more robust datetime code if all local datetime handling happens in the local timezone.

Reply:
Update: I was wrong. It was sufficient to replace tzinfo with `timezone.utc` instead of using `astimezone` which also does conversion. Dealing with time zones is hard. :)

Reply:
Great, this looks better.